File: JAVA/Rotate List.java
public class Solution {
    public ListNode rotateRight(ListNode head, int n) {
        if (head == null) return null;
        int len = 0;
        ListNode node = head;
        while (node != null)
        {
            node = node.next;
            len++;
        }
        n = n % len;
        if (n == 0) return head;
        n = len - n;
        node = head;
        ListNode last = null;
        while (n != 0)
        {
            last = node;
            node = node.next;
            n--;
        }
        last.next = null;
        ListNode end = node;
        while (end.next != null)
        {
            end = end.next;
        }
        end.next = head;
        return node;
    }
}

File: Rotate List.cpp
class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (head == NULL) return NULL;
        ListNode *node = head;
        int len = 1;
        while (node->next != NULL) {
            node = node->next;
            len++;
        }
        k %= len;
        if (k == 0) return head;
        ListNode *end = node;
        node = head;
        for (int i = 0; i < len - k - 1; i++) {
            node = node->next;
        }
        ListNode *newHead = node->next;
        node->next = NULL;
        end->next = head;
        return newHead;
    }
};
